opm/autodiff/NewtonIterationBlackoilInterleaved.cpp

```diff
@@ -48,245 +56,444 @@ namespace Opm
+            if (!result.converged) {
+                OPM_THROW(LinearSolverProblem, "Convergence failure for linear solver.");
+            }
+
+            // Copy solver output to dx.
+            for (int i = 0; i < size; ++i) {
+                for( int p=0, idx = i; p<np; ++p, idx += size ) {
+                    dx(idx) = x[i][p];
                 }
             }
+
+            if ( hasWells ) {
+                // Compute full solution using the eliminated equations.
+                // Recovery in inverse order of elimination.
+                dx = recoverVariable(elim_eqs[1], dx, np);
+                dx = recoverVariable(elim_eqs[0], dx, np);
+            }
+            return dx;
         }
+
+    protected:
+        mutable int iterations_;
+        boost::any parallelInformation_;
+
+        NewtonIterationBlackoilInterleavedParameters parameters_;
+    }; // end NewtonIterationBlackoilInterleavedImpl
+
+
+
+    /// Construct a system solver.
+    NewtonIterationBlackoilInterleaved::NewtonIterationBlackoilInterleaved(const parameter::ParameterGroup& param,
+                                                                           const boost::any& parallelInformation_arg)
+      : newtonIncrement_(),
+        parameters_( param ),
+        parallelInformation_(parallelInformation_arg),
+        iterations_( 0 )
+    {
     }
 
+    namespace detail {
+
+        template< int NP >
+        struct NewtonIncrement
+        {
+            template <class NewtonIncVector>
+            static const NewtonIterationBlackoilInterface&
+            get( NewtonIncVector& newtonIncrements,
+                 const NewtonIterationBlackoilInterleavedParameters& param,
+                 const boost::any& parallelInformation,
+                 const int np )
+            {
+                if( np == NP )
+                {
+                    assert( np < int(newtonIncrements.size()) );
+                    // create NewtonIncrement with fixed np
+                    if( ! newtonIncrements[ NP ] )
+                        newtonIncrements[ NP ].reset( new NewtonIterationBlackoilInterleavedImpl< NP >( param, parallelInformation ) );
+                    return *(newtonIncrements[ NP ]);
+                }
+                else
+                {
+                    return NewtonIncrement< NP-1 >::get(newtonIncrements, param, parallelInformation, np );
+                }
+            }
+        };
 
+        template<>
+        struct NewtonIncrement< 0 >
+        {
+            template <class NewtonIncVector>
+            static const NewtonIterationBlackoilInterface&
+            get( NewtonIncVector& newtonIncrements,
+                 const NewtonIterationBlackoilInterleavedParameters&,
+                 const boost::any&,
+                 const int )
+            {
+                OPM_THROW(std::runtime_error,"NewtonIncrement::get: number of variables not supported yet. Adjust maxNumberEquations appropriately to cover np = " << 0);
```

Comment:
Final (I think) minor detail: we always print "...to cover np = 0" instead of the actual np value.